crates/bevy_ecs/src/world/error.rs

```diff
@@ -55,3 +55,17 @@ pub enum EntityMutableFetchError {
     #[error("The entity with ID {0} was requested mutably more than once")]
     AliasedMutability(Entity),
 }
+
+/// An error that occurs when getting a resource of a given type in a world.
+#[derive(thiserror::Error, Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ResourceFetchError {
+    /// The resource does not exist in the world.
+    #[error("The resource does not exist in the world.")]
+    MissingResource,
+    /// The resource with the given [`ComponentId`] is not initialized in the world.
+    #[error("The resource with ID {0:?} is not initialized in the world.")]
+    NotInitialized(ComponentId),
```

Comment:
The names here feel a little backwards.  I might suggest `NotRegistered` to match the term from `World::register_resource` and `DoesNotExist` to match `EntityDoesNotExistError`.  

```suggestion
    /// The resource has never been initialized or registered with the world.
    #[error("The resource has never been initialized or registered with the world.")]
    NotRegistered,
    /// The resource with the given [`ComponentId`] does not currently exist in the world.
    #[error("The resource with ID {0:?} does not currently exist in the world.")]
    DoesNotExist(ComponentId),
```

We may even want the "Did you forget...?" line from the panic message in `World::resource` to be part of the display string for `NotRegistered`: https://github.com/bevyengine/bevy/blob/81c0900cf23218732c013656669dfdeece2937a8/crates/bevy_ecs/src/world/mod.rs#L1881-L1884

Reply:
Updated it, please check if this is what you had in mind :)

Reply:
> Updated it, please check if this is what you had in mind :)

Yup, that looks good!!